### PR TITLE
Add language filter

### PR DESCRIPTION
**Pull request by jackyho112**
https://github.com/bukinoshita/open-source/issues/1

### CODE REVIEW

File: components/ListItem.js [hunk 1]
```diff
@@ -1,17 +1,35 @@
...
-  <a className="list-item-link" href={props.html_url}>
-    <h3 className="list-item__title">
-      <Truncate lines={2}>{props.title}</Truncate>
-    </h3>
-    <p className="list-item__description">
-      <Truncate lines={3}>{props.body || 'No description on this issue'}</Truncate>
-    </p>
-  </a>
-</li>
+    <div className="row">
```
Comment (bukinoshita) [inline components/ListItem.js]: Add this `div` inside the `<a className="list-item-link" href={props.html_url}>` and remove the class `row` of it

File: components/Footer.js [hunk 2]
```diff
@@ -0,0 +1,19 @@
+import React from 'react'
+
+const Footer = () => (
+  <footer className="footer">
+    <p className="footer__text">
+      Made by
```
Comment (bukinoshita) [inline components/Footer.js]: Missing space between `by` and `Bu` (byBu)

File: components/Footer.js [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+      <a href="https://twitter.com/bukinoshita" className="footer__text--link">
+        Bu Kinoshita
+      </a>
+      . Check this project on
```
Comment (bukinoshita) [inline components/Footer.js]: Missing space between `on` and `Github` (onGithub)

File: pages/index.js [hunk 4]
```diff
@@ -1,20 +1,71 @@
...
+import '../env'
 import 'isomorphic-fetch'
+import React, { Component } from 'react'
 import Head from 'next/head'
-import ListItem from '../components/ListItem';
+import ListItem from '../components/ListItem'
+import FilterButton from '../components/FilterButton'
+import Footer from '../components/Footer'
+import getRepoLanguagesStoreFromIssues from '../services/getRepoLanguagesStoreFromIssues'
+import flatten from 'lodash/flatten'
```
Comment (bukinoshita) [inline pages/index.js]: I think you can import `lodash` like this, give it a try:

```js
import { flatten, uniq, includes, set } from 'lodash'
```

**Review by bukinoshita — CHANGES_REQUESTED** [review]
Awesome work @jackyho112! It's looking really good 🎉 🌮  It's working like a charm! Loved!

I requested some changes on the pull request before merging, but also want to point some/others things here

## Some thought:

### env.js example
You may want to add an example for `env.js` on `README.md` and how to use it. Some people may have problems not knowing how to configure it. (or create an issue to add later)

### Missing space on footer
You've created `components/Footer.js` and now there's no space between `byBu` and `onGithub`. Should be `by Bu` and `on Github`.

### Filter
Button container CSS:
```css
.button__container {
  background-color: #ffffff;
  margin-bottom: 50px;
  padding-left: 15px;
  padding-top: 15px;
  display: flex;
  flex-wrap: wrap;
  justify-content: center;
}
```

Change to:

```css
.button__container {
  background-color: #ffffff;
  margin-bottom: 50px;
  padding-left: 15px;
  padding-top: 15px;
  padding-right: 15px;
  display: flex;
  flex-wrap: wrap;
  justify-content: space-between;
}
```

On `.filter__button` just remove `margin-right: 15px`.

And then add a `.filter__button--fake` to make the alignment right

Check on on `index.js` we have an example:
```html
<ul className="list">
  {issueList}

  <li className="fakeit"/>
  <li className="fakeit"/>
</ul>
```

### List Item

This should be inside the `<a class="list-item-link">` and also remove the class `row` on this div.
```js
<div className="row">
  {
    props.languages.map((language) => (
      <button onClick={() => props.onCategoryFilter(language)} key={`${language}-list-item`}>
        {language}
      </button>
    ))
  }
</div>
```

**Comment by jackyho112** [comment]
@bukinoshita I was going to ask for your feedback but then you already did it 😃 . Thanks for your comments. I will make the changes and then ask for your thought again. Also, the language filters within the list item are just plain buttons for now. I can fix them to match the design once I get the SVGs 👍 .

**Comment by bukinoshita** [comment]
Cool! All the functionality is working flawless as far as I tested 🥇 

I just realized that there's too many programming language LOL

So I quickly mockup this for you
screenshot 2017-02-25 12 19 14

```html
<span style="
  font-size: 12px;
  background-color: #5f357d;
  color: #fff;
  padding: 6px 8px;
  margin-bottom: 10px;
  display: inline-block;
">Javascript</span>
```

By doing that, you may want to increase `.list-item` height
```css
.list-item {
  background-color: #fff;
  flex-basis: 32%;
  margin-bottom: 25px;
  box-shadow: 0 2px 10px rgba(0, 0, 0, .2);
  transition: .15s;
  min-height: 220px; /* it was 200px */
  max-height: 220px;  /* it was 200px */
  overflow: hidden;
}

**Comment by jackyho112** [comment]
@bukinoshita Hello, I made some changes as requested. The only thing I didn't do is to align the filter buttons with fake buttons. I'm not sure how to achieve that with a variable number of filter buttons/languages using the fake button mechanic. I could do that with `margin-right: 10px` on the button and `justify-content: center` on the container. Not sure if that's how you want it to look though. Let me know your thought. I am open to anything 😄 .

Another note is that the performance could drop for having a large number of list items from a small number of list items. One example is changing from the C language filter to the zero language filter. One possible solution is to not use the `Truncate` component. Not sure about the alternative but that's just an idea. 

Let me know if there is anything else you want to give feedback on or if I miss anything. I will love to hear your thoughts on anything. 🥂

**Comment by bukinoshita** [comment]
I just realized something @jackyho112. We should get the repository language for `list-item` from [here](https://api.github.com/repos/bukinoshita/open-source) instead of having an array of languages

```
language: "JavaScript"
```

Some of them has too many languages, I guess would be better to show only Python as language since Python is the primary language of this repo (on this example below). What do you think?
screenshot 2017-02-25 13 52 25

I didn't understand this, can you explain to me what's happening?
> Another note is that the performance could drop for having a large number of list items from a small number of list items. One example is changing from the C language filter to the zero language filter. One possible solution is to not use the Truncate component. Not sure about the alternative but that's just an idea.

And again, thanks! You're doing an amazing work!

**Comment by jackyho112** [comment]
@bukinoshita I believe I am already getting the languages from this API endpoint(I think that's only way to get the information language for a repo). Currently, what I'm doing is get all the repo urls, get all the language urls from these repo urls, fetch all the languages from them and then map them to an object named `repoLanguageStore` in a variable. 

I'm not sure if showing only the one main language is the most user-friendly approach because someone may want to filter the issues with a language that may not be the main language of any repo. Once she gets the list, none of the issues shows the filter language as included, which may seem weird and unexpected. Maybe a 'More' button would be more appropriate. 

What I meant by my comment regarding performance is that loading a list of a large number of list items may take a bit longer than what the user deems responsive. One suggestion is to take off the `Truncate` component and maybe use something else or just implement what you need. But it's actually not a big deal.

Thank you again for the opportunity. I am glad I helped. Let me know if there is anything else you have in mind or any change you want.

**Comment by bukinoshita** [comment]
Got it! I agree with you, let's stick with this then ☺️

I don't want to remove the truncate because we can't control the descriptions of the issues, some of them has large description.

Huge thanks!

**Review by bukinoshita — APPROVED** [review]

**Comment by ulgens** [comment]
Did you release this feature? Current website has no filtering option or language information.

**Comment by bukinoshita** [comment]
Yes @ulgens, this is the last version.

kapture 2017-03-20 at 8 07 30

**Pull request merged**